// tone/Tone/core/clock/TransportEvent.js
import { noOp } from "../util/Interface.js";
/**
 * TransportEvent is an internal class used by {@link TransportClass}
 * to schedule events. Do no invoke this class directly, it is
 * handled from within Tone.Transport.
 */
export class TransportEvent {
    /**
     * Reference to the Transport that created it
     */
    transport;
    /**
     * The unique id of the event
     */
    id = TransportEvent._eventId++;
    /**
     * The time the event starts
     */
    time;
    /**
     * The callback to invoke
     */
    callback;
    /**
     * If the event should be removed after being invoked.
     */
    _once;
    /**
     * The remaining value between the passed in time, and Math.floor(time).
     * This value is later added back when scheduling to get sub-tick precision.
     */
    _remainderTime = 0;
    /**
     * @param transport The transport object which the event belongs to
     */
    constructor(transport, opts) {
        const options = Object.assign(TransportEvent.getDefaults(), opts);
        this.transport = transport;
        this.callback = options.callback;
        this._once = options.once;
        this.time = Math.floor(options.time);
        this._remainderTime = options.time - this.time;
    }
    static getDefaults() {
        return {
            callback: noOp,
            once: false,
            time: 0,
        };
    }
    /**
     * Current ID counter
     */
    static _eventId = 0;
    /**
     * Get the time and remainder time.
     */
    get floatTime() {
        return this.time + this._remainderTime;
    }
    /**
     * Invoke the event callback.
     * @param  time  The AudioContext time in seconds of the event
     */
    invoke(time) {
        if (this.callback) {
            const tickDuration = this.transport.bpm.getDurationOfTicks(1, time);
            this.callback(time + this._remainderTime * tickDuration);
            if (this._once) {
                this.transport.clear(this.id);
            }
        }
    }
    /**
     * Clean up
     */
    dispose() {
        this.callback = undefined;
        return this;
    }
}
