// tone/Tone/signal/GreaterThan.js
import { optionsFromArguments } from "../core/util/Defaults.js";
import { Subtract } from "./Subtract.js";
import { Signal } from "./Signal.js";
import { GreaterThanZero } from "./GreaterThanZero.js";
import { readOnly } from "../core/util/Interface.js";
/**
 * Output 1 if the signal is greater than the value, otherwise outputs 0.
 * can compare two signals or a signal and a number.
 *
 * @example
 * return Tone.Offline(() => {
 * 	const gt = new Tone.GreaterThan(2).toDestination();
 * 	const sig = new Tone.Signal(4).connect(gt);
 * }, 0.1, 1);
 * @category Signal
 */
export class GreaterThan extends Signal {
    name = "GreaterThan";
    override = false;
    input;
    output;
    /**
     * compare that amount to zero after subtracting
     */
    _gtz;
    /**
     * Subtract the value from the input node
     */
    _subtract;
    /**
     * The signal to compare to the incoming signal against.
     * @example
     * return Tone.Offline(() => {
     * 	// change the comparison value
     * 	const gt = new Tone.GreaterThan(1.5).toDestination();
     * 	const signal = new Tone.Signal(1).connect(gt);
     * 	gt.comparator.setValueAtTime(0.5, 0.1);
     * }, 0.5, 1);
     */
    comparator;
    constructor() {
        const options = optionsFromArguments(GreaterThan.getDefaults(), arguments, ["value"]);
        super(options);
        this._subtract = this.input = new Subtract({
            context: this.context,
            value: options.value,
        });
        this._gtz = this.output = new GreaterThanZero({
            context: this.context,
        });
        this.comparator = this._param = this._subtract.subtrahend;
        readOnly(this, "comparator");
        // connect
        this._subtract.connect(this._gtz);
    }
    static getDefaults() {
        return Object.assign(Signal.getDefaults(), {
            value: 0,
        });
    }
    dispose() {
        super.dispose();
        this._gtz.dispose();
        this._subtract.dispose();
        this.comparator.dispose();
        return this;
    }
}
